Mathlib/Algebra/Order/Hom/Quantale.lean

```diff
@@ -0,0 +1,218 @@
+as a function via this coercion.
+
+In the file `Mathlib.Algebra.Order.Hom.Monoid` it is mentioned that there used to be classes:
+`OrderAddMonoidHomClass` etc., but that in #10544 there was a migration from these typeclasses to
+assumptions like `[FunLike F M N] [MonoidHomClass F M N] [OrderHomClass F M N]`.
+We follow that approach here as well, and only define `AddQuantaleHom` without needing
+`AddQuantaleHomClass`.
+
+## Tags
+
+quantale, ordered semigroup, complete lattice
+
+-/
+
+open Function
+
+variable {F α β γ δ : Type*}
+
+section AddQuantale
+
+/-- `α →ₙ+q β` is the type of monotone functions `α → β` that preserve the `AddQuantale`
+structure.
+
+When possible, instead of parametrizing results over `(f : α →ₙ+q β)`,
+you should parametrize over
+`(F : Type*) [FunLike F M N] [AddHomClass F M N] [CompleteLatticeHomClass F M N] (f : F)`. -/
+structure AddQuantaleHom (α β : Type*)
+  [AddSemigroup α] [CompleteLattice α] [AddSemigroup β] [CompleteLattice β]
+  extends α →ₙ+ β, sSupHom α β
+
+/-- Converts an AddQuantaleHom to a sSupHom -/
+add_decl_doc AddQuantaleHom.tosSupHom
+
+/-- Infix notation for `AddQuantaleHom`. -/
+infixr:25 " →ₙ+q " => AddQuantaleHom
+
+-- Instances and lemmas are defined below through `@[to_additive]`.
+
+end AddQuantale
+
+section Quantale
+
+/-- `α →ₙ*q β` is the type of monotone functions `α → β` that preserve the `Quantale`
+structure.
+
+When possible, instead of parametrizing results over `(f : α →ₙ*q β)`,
+you should parametrize over
+`(F : Type*) [FunLike F M N] [MulHomClass F M N] [CompleteLatticeHomClass F M N] (f : F)`. -/
+@[to_additive existing]
+structure QuantaleHom (α β : Type*)
+  [Semigroup α] [CompleteLattice α] [Semigroup β] [CompleteLattice β]
+  extends α →ₙ* β, sSupHom α β
+
+/-- Converts a QuantaleHom to a sSupHom -/
+add_decl_doc QuantaleHom.tosSupHom
+
+/-- Infix notation for `QuantaleHom`. -/
+infixr:25 " →ₙ*q " => QuantaleHom
+
+variable [Semigroup α] [Semigroup β] [CompleteLattice α] [CompleteLattice β] [FunLike F α β]
+
+/-- Turn an element of a type `F` satisfying `MulHomClass F α β` and `sSupHomClass F α β`
+into an actual `QuantaleHom`. This is declared as the default coercion from `F` to `α →ₙ*q β`. -/
+@[to_additive (attr := coe)
+  "Turn an element of a type `F` satisfying `AddHomClass F α β` and `sSupHomClass F α β`
+into an actual `AddQuantaleHom`. This is declared as the default coercion from `F` to `α →ₙ+q β`."]
+def toQuantaleHom [MulHomClass F α β] [sSupHomClass F α β] (f : F) :
+    α →ₙ*q β := { (f : α →ₙ* β) with map_sSup' := sSupHomClass.map_sSup f }
+
+/-- Any type satisfying `QuantaleHomClass` can be cast into `QuantaleHom` via
+  `QuantaleHomClass.toQuantaleHom`. -/
+@[to_additive "Any type satisfying `AddQuantaleHomClass` can be cast into `AddQuantaleHom` via
+  `AddQuantaleHomClass.toAddQuantaleHom`."]
+instance [MulHomClass F α β] [sSupHomClass F α β] : CoeTC F (α →ₙ*q β) :=
+  ⟨toQuantaleHom⟩
```

Comment:
Per what I've said in various places, I now think we shouldn't have this generic coercion but instead have concrete coercions on concrete types...
```suggestion
```